pkg/actuators/machineset/ec2_instance_types.go

```diff
@@ -13,3763 +13,166 @@ limitations under the License.
-		MemoryMb:     262144,
-		GPU:          0,
-		Architecture: "amd64",
-	},
-	"x2iezn.4xlarge": {
-		InstanceType: "x2iezn.4xlarge",
-		VCPU:         16,
-		MemoryMb:     524288,
-		GPU:          0,
-		Architecture: "amd64",
-	},
-	"x2iezn.6xlarge": {
-		InstanceType: "x2iezn.6xlarge",
-		VCPU:         24,
-		MemoryMb:     786432,
-		GPU:          0,
-		Architecture: "amd64",
-	},
-	"x2iezn.8xlarge": {
-		InstanceType: "x2iezn.8xlarge",
-		VCPU:         32,
-		MemoryMb:     1048576,
-		GPU:          0,
-		Architecture: "amd64",
-	},
-	"x2iezn.metal": {
-		InstanceType: "x2iezn.metal",
-		VCPU:         48,
-		MemoryMb:     1572864,
-		GPU:          0,
-		Architecture: "amd64",
-	},
-	"z1d.12xlarge": {
-		InstanceType: "z1d.12xlarge",
-		VCPU:         48,
-		MemoryMb:     393216,
-		GPU:          0,
-		Architecture: "amd64",
-	},
-	"z1d.2xlarge": {
-		InstanceType: "z1d.2xlarge",
-		VCPU:         8,
-		MemoryMb:     65536,
-		GPU:          0,
-		Architecture: "amd64",
-	},
-	"z1d.3xlarge": {
-		InstanceType: "z1d.3xlarge",
-		VCPU:         12,
-		MemoryMb:     98304,
-		GPU:          0,
-		Architecture: "amd64",
-	},
-	"z1d.6xlarge": {
-		InstanceType: "z1d.6xlarge",
-		VCPU:         24,
-		MemoryMb:     196608,
-		GPU:          0,
-		Architecture: "amd64",
-	},
-	"z1d.large": {
-		InstanceType: "z1d.large",
-		VCPU:         2,
-		MemoryMb:     16384,
-		GPU:          0,
-		Architecture: "amd64",
-	},
-	"z1d.metal": {
-		InstanceType: "z1d.metal",
-		VCPU:         48,
-		MemoryMb:     393216,
-		GPU:          0,
-		Architecture: "amd64",
-	},
-	"z1d.xlarge": {
-		InstanceType: "z1d.xlarge",
-		VCPU:         4,
-		MemoryMb:     32768,
-		GPU:          0,
-		Architecture: "amd64",
-	},
+// InstanceTypesCache is a cache for instance type information.
+type InstanceTypesCache interface {
+	GetInstanceType(awsClient awsclient.Client, cacheID string, instanceType string) (InstanceType, error)
+}
+
+// instanceTypesRegion holds cached instance types for specific region and time when it was last updated.
+type instanceTypesRegion struct {
+	instanceTypes map[string]InstanceType
+	lastUpdate    time.Time
+}
+
+// instanceTypesCache holds cached instance types per region. Acess is synchronized via rwmutex.
+type instanceTypesCache struct {
+	cache   map[string]instanceTypesRegion
+	rwmutex sync.RWMutex
+}
+
+// NewInstanceTypesCache creates an empty instance types cache.
+func NewInstanceTypesCache() InstanceTypesCache {
+	cache := &instanceTypesCache{}
+	cache.cache = map[string]instanceTypesRegion{}
+	cache.rwmutex = sync.RWMutex{}
+	return cache
+}
+
+// GetInstanceType retrievees InstanceType from cache by name. If the cache is stale or nil it is refreshed first from the EC2 API.
+// The fetched instance types are specific to the region of the awsClient. Using region name as cacheID is recomended.
+func (i *instanceTypesCache) GetInstanceType(awsClient awsclient.Client, cacheID string, instanceType string) (InstanceType, error) {
+	i.rwmutex.RLock()
```

Comment:
You should probably `defer i.rwmutex.RUnlock()` as I think in at least one place you aren't unlocking it on return in this function

Reply:
Using defer creates a path where you have to Rlock unnecessarily after it fails to refresh just so that defer can unlock. Added the missing RUnlock.